query/src/lib.rs

```diff
@@ -345,6 +351,81 @@ impl Contains {
     }
 }
 
+pub struct StartsWith {
+    array: Arc<dyn Array>,
+}
+
+impl StartsWith {
+    pub fn new(array: Arc<dyn Array>) -> Result<Self> {
+        if array.is_nullable() {
+            return Err(anyhow!(
+                "cannot construct starts_with filter with a nullable array"
+            ));
+        }
+
+        Ok(Self { array })
+    }
+
+    fn starts_with(&self, arr: &dyn Array) -> Result<BooleanArray> {
```

Comment:
I think starts_with will be less performant than contains because we can't use the hashmap optimization, it needs to run through every tuple.

Watching the CMU classes, this is exactly the kind of implementation that can have lots of optimizations to be made and that I think we shouldn't implement it manually but rather use a library. I even think that arrow might have a filter function.

Reply:
Hashmap only optimizes the query set, you have to go through the entire thing anyways. So you can just use a btreemap for this similar to how contains uses hashmap

Reply:
You can use this https://docs.rs/arrow/latest/arrow/compute/kernels/comparison/fn.starts_with.html

The reason I don't want to use datafusion and polars is that I tried them several times and you hit too many issues like the operator not supporting some datatype or running slow for some reason etc. 

The things we are implementing are very basic so using whole query engine isn't worth it